
## ---------------------------------------------------------------------------------------------------------------------
#
#                                           Bioinformatics Research Group
#                                             http://biorg.cis.fiu.edu/
#                                         Florida International University
#
#
#
#   This software is a "Camilo Valdes Work" under the terms of the United States Copyright Act.
#   Please cite the author(s) in any work or product based on this material.
#
#   OBJECTIVE:
#	  The purpose of this program is to create a plot that will visualize the input number of reads against the output
#   number of alignments generated by Bowtie2.
#
#   NOTES:
#           THE SCRIPT IS DEPRECATED.
#
#   DEPENDENCIES:
#
#       • ggplot2
#       • RColorBrewer
#       • reshape2
#       • gridExtra
#       • grid
#
#   The above libraries & modules are required.
#
#   AUTHOR:	Camilo Valdes (cvalde03@fiu.edu)
#           Bioinformatics Research Group,
#           School of Computing and Information Sciences,
#           Florida International University (FIU)
#
#
# ---------------------------------------------------------------------------------------------------------------------


#
# 	Import any necessary system or 3rd-party libraries & frameworks here
#
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(gridExtra)
library(grid)

dyn.load(paste("RPluMA", .Platform$dynlib.ext, sep=""))
source("RPluMA.R")

input <- function(inputfile) {
  parameters <<- read.table(inputfile, as.is=T);
  rownames(parameters) <<- parameters[,1];
    pfix = prefix()
  if (length(pfix) != 0) {
     pfix <- paste(pfix, "/", sep="")
  }
  #labelsForSamples = c('Nasopharynx', 'Throat', 'S.Plaque', 'Nasopharynx', 'Throat', 'S.Plaque')
  #dataValues = c('0.066', '3.3', '45', '4', '15', '126')
  #typeForDataValues = c('input', 'input', 'input', 'output', 'output', 'output')
  #numeric_vars <- readLines(paste(pfix, parameters["features", 2], sep="/"))  
  labelsForSamples <<- readLines(paste(pfix, parameters["labels", 2], sep="/"))
  dataValues <<- readLines(paste(pfix, parameters["data", 2], sep="/"))
  typeForDataValues <<- readLines(paste(pfix, parameters["inputs", 2], sep="/"))
}

run <- function() {}

output <- function(outputfile) {
pdf(paste(outputfile, "results.pdf", sep='/'))
	# 	Set some options to alter the default behavior of the REPL.
options( width=512, digits=15, warn=1 )


#
#	Custom Colors for Color Blind folks
#
# 	The palette with black:
cbbPalette_black = c( "#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7" )

# 	The palette with grey:
cbPalette_grey = c( "#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7" )


# ------------------------------------------------- Project Setup -----------------------------------------------------

print( paste( "[", format(Sys.time(), "%m/%d/%y %H:%M:%S"),"] ", sep="") )
print( paste( "[", format(Sys.time(), "%m/%d/%y %H:%M:%S"),"] R Starting... ", sep="") )
print( paste( "[", format(Sys.time(), "%m/%d/%y %H:%M:%S"),"] ", sep="") )


#working_directory="/path/to/working/directory/visualization"
#setwd( file.path( working_directory ) )

figures_base_dir=paste(outputfile , sep="")
pdf_output_dir=paste( outputfile, sep="" )
png_output_dir=paste( outputfile, sep="" )

#dir.create( file.path( pdf_output_dir ), showWarnings=FALSE, recursive=TRUE )
#dir.create( file.path( png_output_dir ), showWarnings=FALSE, recursive=TRUE )


# ------------------------------------------------------ Main ---------------------------------------------------------

#   Headers for the dataframe
columnNamesForData = c('Sample', 'DataValue', 'Type')

#
#   Data. Reads are in Millions, and output is in Gigabytes.
#   Previously obtained, and hard-wired for testing.
#

#   Create a dataframe for plotting
dataFrameWithReadsAndOutput = data.frame(labelsForSamples, dataValues, typeForDataValues, stringsAsFactors=FALSE)
colnames(dataFrameWithReadsAndOutput) = columnNamesForData

#	Label column should be "Factor" for ggplot later on.
dataFrameWithReadsAndOutput[,1:1] = sapply(dataFrameWithReadsAndOutput[,1:1], as.factor)
#	Remaining columns should be "Numeric" as they are data values.
dataFrameWithReadsAndOutput[,2:2] = sapply(dataFrameWithReadsAndOutput[,2:2], as.numeric)

str(dataFrameWithReadsAndOutput)
dataFrameWithReadsAndOutput

#
#   Plot the data using a double sided Histogram
#
plotForData = ggplot( dataFrameWithReadsAndOutput, aes( x=Sample))


outputPlot = plotForData + geom_bar( data=subset(dataFrameWithReadsAndOutput, Type=='output'),
						aes(y=DataValue), fill="brown2", stat="identity") +
						geom_text(aes(y=DataValue, label=c('4', '15', '126', '', '', '')), hjust=-2) +
						scale_fill_manual(values="Spectral") +
						scale_y_continuous('Gigabytes', limits=c(0, 150)) +
  						theme(legend.position = 'none',
        						axis.title.y = element_blank(),
        						#axis.text.y = element_text(size=8),
        						plot.title = element_text(size=11.5),
        						plot.margin=unit(c(0.1,0.2,0.1,-.1),"cm"),
        						axis.ticks.y = element_blank(),
        						axis.text.y = theme_bw()$axis.text.y) +
  						ggtitle("Output Alignments") +
  						coord_flip()


inputPlot = plotForData + geom_bar( data=subset(dataFrameWithReadsAndOutput, Type=='input'),
						aes(y=DataValue), fill="chartreuse4", stat="identity") +
						geom_text(aes(y=DataValue, label=c('0.066', '3.3', '45', '', '', '')), hjust=1.25) +
						scale_y_continuous('Millions of Reads', limits=c(150, 0), trans='reverse') +
						theme(legend.position = 'none',
        						axis.text.y = element_blank(),
        						axis.ticks.y = element_blank(),
        						plot.title = element_text(size=10.5),
        						plot.margin=unit(c(0.1,0,0.1,0.05),"cm")) +
						ggtitle("Input Reads") +
						coord_flip() +
    					xlab("HMP Samples")


grid.arrange( inputPlot, outputPlot, widths=c(0.45,0.55), ncol=2 )


# ------------------------------------------------------ END ---------------------------------------------------------

print( paste( "[", format(Sys.time(), "%m/%d/%y %H:%M:%S"),"] ", sep="") )
print( paste( "[", format(Sys.time(), "%m/%d/%y %H:%M:%S"),"] Done.", sep="") )
print( paste( "[", format(Sys.time(), "%m/%d/%y %H:%M:%S"),"] ", sep="") )
}
